### PR TITLE
Implement ScyllaDb as a `KeyValueStoreClient`

### PR DESCRIPTION
**Pull request by MathieuDutSik**
# Motivation

The `ScyllaDb` database is a client that we want to support. A big reason for supporting it is that `DynamoDb` is available only on AWS and we are making first step of supporting GCP which could create a problematic situation. A second reason is that it is faster and cheaper than `DynamoDb` (they claim 7x cheaper which is likely overblown but still likely cheaper in realistic scenarios). This PR is the first one for supporting it.

# Solution

A client is created that implements the `KeyValueStoreClient`. A constructor `new` is created that allows creating it using an `uri` and a `namespace` (which is a primary key in database parlance). There is a feature `scylladb` in accordance with the practice 

A corresponding `create_scylladb_test_client` is written that allows creating a client and making the needed tests. The obtained client passes the `test_readings_scylladb` and `test_writings_scylladb` which likely means that everything is ok. 

# Test Plan

The testing is done via following commands
* Pull the docker via `docker pull scylladb/scylla` 
* Run the docker via `docker run --name my_scylla_container -d -p 9042:9042 scylladb/scylla`
* Then test via `cargo test test_readings_scylla_db` and `cargo test test_writings_scylla_db`.

Next steps:
* Making the tests supported in CI.
* Strengthen it by implementing journaling.
* Make it part of all the other tests on par with memory, RocksDb and DynamoDb.

### CODE REVIEW

File: linera-views/src/scylla_db.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+/// The creation of a ScyllaDb client that can be used for accessing it.
+/// The `Vec<u8>`is a primary key.
+#[derive(Clone)]
+pub struct ScyllaDbClient {
+    client: Arc<RwLock<ScyllaDbClientPair>>,
+}
+
+/// The error type for [`ScyllaDbClient`]
+#[derive(Error, Debug)]
+pub enum ScylladbContextError {
```
Comment (ma2bd) [inline linera-views/src/scylla_db.rs:46]: ScyllaDb

File: linera-views/src/scylla_db.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let client = self.client.read().await;
+        find_key_values_by_prefix_internal(&client, key_prefix.to_vec()).await
+    }
+
+    async fn write_batch(&self, batch: Batch, _base_key: &[u8]) -> Result<(), Self::Error> {
+        let client = self.client.read().await;
+        write_batch_internal(&client, batch).await
```
Comment (ma2bd) [inline linera-views/src/scylla_db.rs:252]: What is the point of moving out the code into a top-level function?

File: linera-views/tests/operations_tests.rs [hunk 3]
```diff
@@ -113,7 +116,7 @@ async fn test_readings_vec<OP: KeyValueStoreClient + Sync>(
 
 fn get_random_key_value_vec1(len_value: usize) -> Vec<(Vec<u8>, Vec<u8>)> {
     let key_prefix = vec![0];
-    let n = 1000;
+    let n = 100;
```
Comment (ma2bd) [inline linera-views/tests/operations_tests.rs:119]: Is it this a performance issue?

File: linera-views/tests/operations_tests.rs [hunk 4]
```diff
@@ -170,6 +173,15 @@ async fn test_readings_dynamodb() {
     }
 }
 
+#[cfg(feature = "scylladb")]
+#[tokio::test]
+async fn test_readings_scylladb() {
```
Comment (ma2bd) [inline linera-views/tests/operations_tests.rs:178]: scylla_db?

File: linera-views/src/scylla_db.rs (hunk 2 above)
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:252]: Ok, I put those functions in the `ScyllaDbClient`.

File: linera-views/tests/operations_tests.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-views/tests/operations_tests.rs:119]: Yes, the test test for `n=1000` takes 165s while the test for `n=100` takes 25s.

File: linera-views/src/scylla_db.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:46]: Ok.

File: linera-views/tests/operations_tests.rs (hunk 4 above)
Comment (MathieuDutSik) [reply linera-views/tests/operations_tests.rs]: Ok.

File: linera-views/src/scylla_db.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This provides a KeyValueStoreClient for th ScyllaDB database.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:4]: Done.

File: linera-views/src/scylla_db.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+//! * Make it part of the CI
+//!
+//! The way to make the test work is the following:
+//! * docker pull scylladb/scylla
+//! * docker run --name my_scylla_container -d -p 9042:9042 scylladb/scylla
+//! * cargo test test_readings_scylladb --features scylladb -- --nocapture
+//!
+//! When running the test I actually obtain some error "No connections in the
+//! pool; last connection failed with: Invalid message: Frame error: early eof"
+//! which I assume are network error on my computer.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:21]: Maybe I should make the CI works for the PR to be merged?

[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-views/src/scylla_db.rs:21]: Ideally yes. You can also write a (convincing) "test plan" section in the summary of this task.

File: linera-views/tests/operations_tests.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-views/tests/operations_tests.rs]: Note that the `DynamoDb` takes 360s for `n=1000`. So, we still have improvement.

File: linera-views/src/scylla_db.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs]: Ok, added to the section Test Plan.

File: linera-views/src/lib.rs [hunk 7]
```diff
@@ -54,6 +54,11 @@ pub mod key_value_store_view;
 #[cfg(not(target_arch = "wasm32"))]
 pub mod hashable_wrapper;
 
+/// Helper definitions for Scylladb storage.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/lib.rs:104]: Ok, I renamed it.
In another PR, I will remove those comments of the include (in accordance with other `linera-*` directories).

File: linera-views/src/scylla_db.rs [hunk 8]
```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This provides a KeyValueStoreClient for th ScyllaDB database.
+//! The code is functional but some aspects are missing.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:5]: I added a bunch of corresponding issues on GitHub.

File: linera-views/src/scylla_db.rs [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+};
+use async_lock::Semaphore;
+use async_trait::async_trait;
+use scylla::{IntoTypedRows, Session, SessionBuilder};
+use std::{ops::Deref, sync::Arc};
+use thiserror::Error;
+
+type ScyllaDbClientPair = (Session, Vec<u8>);
+
+const MAX_CONNECTIONS: usize = 10;
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-views/src/scylla_db.rs:29]: In production, I'd rather try to play with configurations here: https://docs.rs/scylla/latest/scylla/transport/session_builder/struct.GenericSessionBuilder.html

Ideally the only semaphore would be for the number of DB tests run in parallel (see rocksdb).

**Comment by ma2bd** [comment]
Please fix the linter. Is the dynamodb error related?

File: linera-views/src/scylla_db.rs [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+/// The creation of a ScyllaDb client that can be used for accessing it.
+/// The `Vec<u8>`is a primary key.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:32]: This was wrongly located, I corrected.

File: linera-views/src/scylla_db.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/scylla_db.rs:29]: Yes, there will be a dynamic configuration after my work on PR 931 gets merged.

File: linera-views/src/common.rs [hunk 11]
```diff
@@ -47,17 +47,25 @@ pub const MIN_VIEW_TAG: u8 = 1;
...
             upper_bound[i] += 1;
-            return Excluded(upper_bound);
+            return Some(upper_bound);
         }
     }
-    Unbounded
+    None
+}
+
+///
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/common.rs:63]: I added the needed documentation.

**Comment by ma2bd** [comment]
Can we rename MAX_CONNECTION into MAX_CONCURRENT_QUERIES or something equivalent?

**Review by ma2bd — APPROVED** [review]

**Pull request merged**